File: Python/03. Understanding Class & Object/practice_07.py
"""
Practice 07: School Homework System (OOAD)
Task: Model a school homework system where teachers create homework,
      students submit answers, and teachers grade the submissions.

How to run:
  python practice_07.py

Key Concepts:
  - Multiple classes collaborating (Teacher, Student, Homework, Submission)
  - Methods that create and return objects
  - State tracking (graded/ungraded)
"""


class Homework:
    """Homework represents an assignment created by a teacher."""

    def __init__(self, title: str, subject: str, due_date: str, description: str, max_score: int):
        """Creates a new Homework assignment."""
        self.title = title
        self.subject = subject
        self.due_date = due_date
        self.description = description
        self.max_score = max_score

    def show_info(self):
        """Prints the homework details."""
        print(f"  Homework    : {self.title}")
        print(f"  Subject     : {self.subject}")
        print(f"  Due Date    : {self.due_date}")
        print(f"  Description : {self.description}")
        print(f"  Max Score   : {self.max_score}")
        print()


class Submission:
    """Submission represents a student's submitted answer for a homework."""

    def __init__(self, student, homework: Homework, answer: str):
        """Creates a new Submission."""
        self.student = student
        self.homework = homework
        self.answer = answer
        self.score = 0
        self.is_graded = False

    def show_info(self):
        """Prints the submission details."""
        print(f"  Student     : {self.student.name}")
        print(f"  Homework    : {self.homework.title}")
        print(f"  Answer      : {self.answer}")
        if self.is_graded:
            print(f"  Score       : {self.score} / {self.homework.max_score}")
        else:
            print("  Score       : Not graded yet")
        print()


class Teacher:
    """Teacher represents a teacher who creates homework and grades submissions."""

    def __init__(self, name: str, subject: str):
        """Creates a new Teacher."""
        self.name = name
        self.subject = subject

    def create_homework(self, title: str, due_date: str, description: str, max_score: int) -> Homework:
        """Creates a new Homework assignment."""
        print(f'  [OK] {self.name} created homework: "{title}"')
        return Homework(title, self.subject, due_date, description, max_score)

    def grade_submission(self, sub: Submission, score: int):
        """Assigns a score to a student's submission."""
        if score < 0 or score > sub.homework.max_score:
            print(f"  [Error] Score must be between 0 and {sub.homework.max_score}.")
            return
        sub.score = score
        sub.is_graded = True
        print(f"  [OK] {self.name} graded {sub.student.name}'s submission: "
              f"{score} / {sub.homework.max_score}")

    def show_info(self):
        """Prints teacher details."""
        print(f"  Teacher: {self.name} | Subject: {self.subject}")


class Student:
    """Student represents a student who submits homework."""

    def __init__(self, name: str, grade: str):
        """Creates a new Student."""
        self.name = name
        self.grade = grade

    def submit_homework(self, hw: Homework, answer: str) -> Submission:
        """Creates a submission for the given homework."""
        print(f'  [OK] {self.name} submitted answer for "{hw.title}"')
        return Submission(self, hw, answer)

    def show_info(self):
        """Prints student details."""
        print(f"  Student: {self.name} | Grade: {self.grade}")


def main():
    # --- Create teacher and students ---
    teacher = Teacher("Mr. Tiemoon", "Mathematics")
    student1 = Student("Tareq", "Grade 10")
    student2 = Student("Afsana", "Grade 10")

    print("=== Participants ===")
    teacher.show_info()
    student1.show_info()
    student2.show_info()
    print()

    # --- Teacher creates homework ---
    print("=== Creating Homework ===")
    hw = teacher.create_homework(
        "Quadratic Equations",
        "2026-03-15",
        "Solve 10 quadratic equations using the quadratic formula.",
        100,
    )
    print()

    print("=== Homework Details ===")
    hw.show_info()

    # --- Students submit homework ---
    print("=== Student Submissions ===")
    sub1 = student1.submit_homework(hw, "x = (-b +/- sqrt(b^2-4ac)) / 2a ... [10 solutions]")
    sub2 = student2.submit_homework(hw, "Used factoring method ... [10 solutions]")
    print()

    # --- Teacher grades submissions ---
    print("=== Grading ===")
    teacher.grade_submission(sub1, 92)
    teacher.grade_submission(sub2, 85)
    print()

    # --- Show final results ---
    print("=== Submission Results ===")
    sub1.show_info()
    sub2.show_info()


if __name__ == "__main__":
    main()

File: Java/03. Understanding Class & Object/Practice07.java
/**
 * Practice 07: School Homework System (OOAD)
 * Task: Model a school homework system where teachers create homework,
 *       students submit answers, and teachers grade the submissions.
 *
 * How to compile and run:
 *   javac Practice07.java
 *   java Practice07
 *
 * Key Concepts:
 *   - Multiple classes collaborating (Teacher, Student, Homework, Submission)
 *   - Methods that create and return objects
 *   - State tracking (graded/ungraded)
 */

public class Practice07 {

    /** Homework represents an assignment created by a teacher. */
    static class Homework {
        private String title;
        private String subject;
        private String dueDate;
        private String description;
        private int maxScore;

        /** Creates a new Homework assignment. */
        public Homework(String title, String subject, String dueDate, String description, int maxScore) {
            this.title = title;
            this.subject = subject;
            this.dueDate = dueDate;
            this.description = description;
            this.maxScore = maxScore;
        }

        /** Prints the homework details. */
        public void showInfo() {
            System.out.printf("  Homework    : %s%n", title);
            System.out.printf("  Subject     : %s%n", subject);
            System.out.printf("  Due Date    : %s%n", dueDate);
            System.out.printf("  Description : %s%n", description);
            System.out.printf("  Max Score   : %d%n", maxScore);
            System.out.println();
        }
    }

    /** Submission represents a student's submitted answer for a homework. */
    static class Submission {
        private Student student;
        private Homework homework;
        private String answer;
        private int score;
        private boolean isGraded;

        /** Creates a new Submission. */
        public Submission(Student student, Homework homework, String answer) {
            this.student = student;
            this.homework = homework;
            this.answer = answer;
            this.score = 0;
            this.isGraded = false;
        }

        /** Prints the submission details. */
        public void showInfo() {
            System.out.printf("  Student     : %s%n", student.name);
            System.out.printf("  Homework    : %s%n", homework.title);
            System.out.printf("  Answer      : %s%n", answer);
            if (isGraded) {
                System.out.printf("  Score       : %d / %d%n", score, homework.maxScore);
            } else {
                System.out.printf("  Score       : Not graded yet%n");
            }
            System.out.println();
        }
    }

    /** Teacher represents a teacher who creates homework and grades submissions. */
    static class Teacher {
        private String name;
        private String subject;

        /** Creates a new Teacher. */
        public Teacher(String name, String subject) {
            this.name = name;
            this.subject = subject;
        }

        /** Creates a new Homework assignment. */
        public Homework createHomework(String title, String dueDate, String description, int maxScore) {
            System.out.printf("  [OK] %s created homework: \"%s\"%n", name, title);
            return new Homework(title, subject, dueDate, description, maxScore);
        }

        /** Assigns a score to a student's submission. */
        public void gradeSubmission(Submission sub, int score) {
            if (score < 0 || score > sub.homework.maxScore) {
                System.out.printf("  [Error] Score must be between 0 and %d.%n", sub.homework.maxScore);
                return;
            }
            sub.score = score;
            sub.isGraded = true;
            System.out.printf("  [OK] %s graded %s's submission: %d / %d%n",
                    name, sub.student.name, score, sub.homework.maxScore);
        }

        /** Prints teacher details. */
        public void showInfo() {
            System.out.printf("  Teacher: %s | Subject: %s%n", name, subject);
        }
    }

    /** Student represents a student who submits homework. */
    static class Student {
        private String name;
        private String grade;

        /** Creates a new Student. */
        public Student(String name, String grade) {
            this.name = name;
            this.grade = grade;
        }

        /** Creates a submission for the given homework. */
        public Submission submitHomework(Homework hw, String answer) {
            System.out.printf("  [OK] %s submitted answer for \"%s\"%n", name, hw.title);
            return new Submission(this, hw, answer);
        }

        /** Prints student details. */
        public void showInfo() {
            System.out.printf("  Student: %s | Grade: %s%n", name, grade);
        }
    }

    public static void main(String[] args) {
        // --- Create teacher and students ---
        Teacher teacher = new Teacher("Mr. Tiemoon", "Mathematics");
        Student student1 = new Student("Tareq", "Grade 10");
        Student student2 = new Student("Afsana", "Grade 10");

        System.out.println("=== Participants ===");
        teacher.showInfo();
        student1.showInfo();
        student2.showInfo();
        System.out.println();

        // --- Teacher creates homework ---
        System.out.println("=== Creating Homework ===");
        Homework hw = teacher.createHomework(
                "Quadratic Equations",
                "2026-03-15",
                "Solve 10 quadratic equations using the quadratic formula.",
                100
        );
        System.out.println();

        System.out.println("=== Homework Details ===");
        hw.showInfo();

        // --- Students submit homework ---
        System.out.println("=== Student Submissions ===");
        Submission sub1 = student1.submitHomework(hw, "x = (-b +/- sqrt(b^2-4ac)) / 2a ... [10 solutions]");
        Submission sub2 = student2.submitHomework(hw, "Used factoring method ... [10 solutions]");
        System.out.println();

        // --- Teacher grades submissions ---
        System.out.println("=== Grading ===");
        teacher.gradeSubmission(sub1, 92);
        teacher.gradeSubmission(sub2, 85);
        System.out.println();

        // --- Show final results ---
        System.out.println("=== Submission Results ===");
        sub1.showInfo();
        sub2.showInfo();
    }
}
